sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumns.scala

```diff
@@ -15,139 +15,385 @@
-   * "b" column retains its original value of "5 + 5".
+   * Example:
+   * CREATE TABLE T(a INT DEFAULT 4, b INT NOT NULL DEFAULT 5);
+   * INSERT INTO T VALUES (1, 2);
+   * INSERT INTO T VALUES (1, DEFAULT);
+   * INSERT INTO T VALUES (DEFAULT, 6);
+   * SELECT * FROM T;
+   * (1, 2)
+   * (1, 5)
+   * (4, 6)
    *
-   * The reason for constant-folding the EXISTS_DEFAULT is to make the end-user visible behavior the
-   * same, after executing an ALTER TABLE ADD COLUMNS command with DEFAULT value, as if the system
-   * had performed an exhaustive backfill of the provided value to all previously existing rows in
-   * the table instead. We choose to avoid doing such a backfill because it would be a
-   * time-consuming and costly operation. Instead, we elect to store the EXISTS_DEFAULT in the
-   * column metadata for future reference when querying data out of the data source. In turn, each
-   * data source then takes responsibility to provide the constant-folded value in the
-   * EXISTS_DEFAULT metadata for such columns where the value is not present in storage.
-   *
-   * @param analyzer used for analyzing the result of parsing the column expression stored as text.
-   * @param tableSchema represents the names and types of the columns of the statement to process.
-   * @param statementType name of the statement being processed, such as INSERT; useful for errors.
-   * @return a copy of `tableSchema` with field metadata updated with the constant-folded values.
+   * @param analyzer analyzer to use for processing DEFAULT values stored as text.
+   * @param catalog  the catalog to use for looking up the schema of INSERT INTO table objects.
+   * @param insert   the enclosing INSERT statement for which this rule is processing the query, if
+   *                 any.
    */
-  def constantFoldCurrentDefaultsToExistDefaults(
+  case class ResolveDefaultColumns(
       analyzer: Analyzer,
-      tableSchema: StructType,
-      statementType: String): StructType = {
-    if (!SQLConf.get.enableDefaultColumns) {
-      return tableSchema
+      catalog: SessionCatalog,
+      insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+      (_ => SQLConf.get.enableDefaultColumns), ruleId) {
+      case i@InsertIntoStatement(_, _, _, _, _, _)
+        if i.query.collectFirst { case u: UnresolvedInlineTable => u }.isDefined =>
+        // Create a helper instance of this same rule with the `insert` argument populated as `i`.
+        // Then recursively apply it on the `query` of `i` to transform the result. This recursive
+        // application lets the below case matching against `UnresolvedInlineTable` to trigger
+        // anywhere that operator may appear in the descendants of `i.query`.
+        val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
```

Comment:
I am thinking about setting a TreeNodeTag with `i` so that we don't need the recursion here.

Reply:
I looked into this, we can do even simpler by just keeping a `var` to track the enclosing `INSERT INTO` command, once we find one. Then we don't need recursion anymore and we can just check the `var` later once we find a `VALUES` or `SELECT` list whose `DEFAULT` references need to be processed.